### DISCUSSION TITLE
For comprehension 'copy' semantics?

### DISCUSSION THREAD

**Discussion by odipar** [cue-lang/cue#1438] [General] [answered]
OK, tried this:

```
X: {
 a: number
 b: a + 1
 c: b + 2
 d: a | 2
}

Y: {
 for k,v in X { "\(k)": v }
}
```

Expected Y to be equal to X?

but got:
```
X: {
	a: number
	b: a + 1
	c: b + 2
	d: number | 2
}
Y: {
	a: number
	b: v
	c: v
	d: number | 2
}
```
Used Cue 0.4.0

See also: https://cuelang.org/play/?id=HT8kx7IjfU8#cue@export@cue

**Comment by xinau** [comment]
To me this mainly feels like a formatting bug, when `X.a` is set a concrete value both structs converge to the same values. I.e.,

```cue
X: {
 a: number
 b: a + 1
 c: b + 2
 d: a | 2
}

Y: {
 for k,v in X { "\(k)": v }
}

X: a: 4
```

results in

```
X: {
    a: 4
    b: 5
    c: 7
    d: 4 | 2
}
Y: {
    a: 4
    b: 5
    c: 7
    d: 4 | 2
}
```

**Reply by odipar** [reply]
I think it is more about eager versus lazy evaluation. It looks like copying a value (expression?) with the comprehension also tries to eagerly evaluate the expression (i.e: `b: a + 1`) but somehow fails and throws an 'error': in this case the identifier `v`? That's really weird behaviour I think.
I understand that in the concrete value case, everything works out perfectly well, but that's because then there is no difference between eager and lazy evaluation.

I would expect the expressions to be copied at verbatim, and may be after that do an evaluation.

**Reply by odipar** [reply]
Also checked with cue version v0.4.1-beta.6 darwin/amd64 with similar result

**Comment by verdverm** [comment]
Have you tried both examples with the cue CLI?

It could be specific to the playground

**Reply by xinau** [reply]
I'm able to confirm that this happens with the CLI as well, when using `eval` (haven't tested other commands).

**Reply by odipar** [reply]
yes, version v0.4.1-beta.6 is the CLI one (as the playground is v0.4.0)

**Comment by odipar** [comment, marked as answer]
btw. i'm ok with CUE to return bottom if the value isn't concrete (although I would rather have a copy of the expression). But returning `v` is a bug I think.

**Comment by odipar** [comment]
This is also weird, but with eager evaluation understandable 
```
X: {
 a: int | *1
 b: a + 5
}

Y: {
 for k,v in X { (k): v }
}

Xa: X & { a: 2 }
Ya: Y & { a: 2 }
```

Result:
```
X: {
    a: 1
    b: 6
}
Y: {
    a: 1
    b: 6
}
Xa: {
    a: 2
    b: 7
}
Ya: {
    a: 2
    b: 6
}
```
shouldn't Xa and Ya be the same?
